### PR TITLE
DEPR: allowing subclass-specific keywords in pd.Index.__new__

### LINKED ISSUE #14093: ERR: stricter checks on Index construction when tz is passed
from [comment](https://github.com/pydata/pandas/pull/13981#issuecomment-239846027) on #13981 

```
In [1]: pd.Index([datetime.datetime(2012,1,1), datetime.datetime(2012,1, 2)], tz="Europe/Brussels")
Out[1]: DatetimeIndex(['2012-01-01 00:00:00+01:00', '2012-01-02 00:00:00+01:00'], dtype='datetime64[ns, Europe/Brussels]', freq=None)

In [2]: pd.Index([1, 2], tz="Europe/Brussels")
Out[2]: DatetimeIndex(['1970-01-01 00:00:00+01:00', '1970-01-01 00:00:00+01:00'], dtype='datetime64[ns, Europe/Brussels]', freq=None)

In [3]: pd.Index(['20120102', '20120103'], tz="Europe/Brussels")
Out[3]: DatetimeIndex(['2012-01-02 00:00:00+01:00', '2012-01-03 00:00:00+01:00'], dtype='datetime64[ns, Europe/Brussels]', freq=None)
```

[1] is fine/expected
[2] should be disallowed (e.g. integer construction)
[3] is ok, maybe should disallow to be consistent.

I am not sure what this would break. But we don't necessarily want to force datetime construction on a generic `Index` (as opposed to using `.to_datetime` or a constructor, e.g. `pd.date_range`). This is a bit too much inference.

So the rule would be that it has to be actual datetimes/Timestamps, if tz is provided (which is a form of dtype).

### LINKED ISSUE #21311: BUG/API: Index constructor does not enforce specified dtype
#### Code Sample, a copy-pastable example if possible

Manually specifying a dtype does not garantuee the output is in that dtype. Eg with Series if incompatible data is passed, an error is raised, while for Index it just silently outputs another dtype:

```python
In [11]: pd.Series(['a', 'b', 'c'], dtype='int64')
...
ValueError: invalid literal for int() with base 10: 'a'

In [12]: pd.Index(['a', 'b', 'c'], dtype='int64')
Out[12]: Index(['a', 'b', 'c'], dtype='object')

```

### LINKED ISSUE #22315: Index() constructor is inconsistent in detecting subclass-specific keyword arguments
#### Code Sample, a copy-pastable example if possible

```python
In [2]: pd.Index(['20000101', '20000102'])
Out[2]: Index(['20000101', '20000102'], dtype='object')

In [3]: pd.Index(['20000101', '20000102'], freq='D')
Out[3]: Index(['20000101', '20000102'], dtype='object')

In [4]: pd.Index(['20000101', '20000102'], tz='utc')
Out[4]: DatetimeIndex(['2000-01-01 00:00:00+00:00', '2000-01-02 00:00:00+00:00'], dtype='datetime64[ns, UTC]', freq=None)
```

#### Problem description

I see no obvious reason why ``tz`` should trigger the interpretation of data as datetime, and ``freq`` shouldn't. And I think the only viable option is to refrain entirely from using kwargs to do inference - unless we are committed to support them all, which could be a pain.

From the point of view of users, the only change would be to replace
```
pd.Index(['20000101', '20000102'], tz='utc')
```
with
```
pd.Index(['20000101', '20000102'], tz='utc', dtype='datetime64')
```

#### Expected Output

``Out[3]`` and ``Out[4]`` should be instances of the same class.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: eb0ac5437f9fee2939b3c3c4dab3ed745b5a9c38
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-6-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.24.0.dev0+460.geb0ac5437
pytest: 3.5.0
pip: 9.0.1
setuptools: 39.2.0
Cython: 0.28.4
numpy: 1.14.3
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.2.2.post1634.dev0+ge8120cf6d
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1
gcsfs: None

</details>

### LINKED ISSUE #26974: Index constructor doesn't validate kwargs
```python
In [4]: pd.Index([1, 2], abc=1)                                                                                                                                                                                                                                                                                                                                           
Out[4]: Int64Index([1, 2], dtype='int64')
```

This should raise, since `abc` isn't a valid keyword.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #21311, closes #14093, closes #22315, closes #26974
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexing/test_coercion.py [hunk 1]
```diff
@@ -506,7 +506,9 @@ def test_insert_index_period(self, insert, coerced_val, coerced_dtype):
         else:
             msg = r"Unexpected keyword arguments {'freq'}"
             with pytest.raises(TypeError, match=msg):
-                pd.Index(data, freq="M")
+                with tm.assert_produces_warning(FutureWarning):
+                    # passing keywords ot pd.Index
```
Comment (arw2019) [inline pandas/tests/indexing/test_coercion.py:510]: typo in comment

File: pandas/core/resample.py [hunk 2]
```diff
@@ -1944,6 +1944,8 @@ def _asfreq_compat(index, freq):
     new_index: Index
     if isinstance(index, PeriodIndex):
         new_index = index.asfreq(freq=freq)
+    elif index.dtype.kind == "M":
+        new_index = DatetimeIndex([], dtype=index.dtype, freq=freq, name=index.name)
     else:
-        new_index = Index([], dtype=index.dtype, freq=freq, name=index.name)
+        new_index = TimedeltaIndex([], dtype=index.dtype, freq=freq, name=index.name)
```
Comment (jreback) [inline pandas/core/resample.py:1950]: can you use another isinstance and then raise on else?

File: pandas/tests/indexes/base_class/test_constructors.py [hunk 3]
```diff
@@ -29,7 +30,8 @@ def test_construction_list_mixed_tuples(self, index_vals):
     def test_constructor_wrong_kwargs(self):
         # GH #19348
         with pytest.raises(TypeError, match="Unexpected keyword arguments {'foo'}"):
-            Index([], foo="bar")
+            with tm.assert_produces_warning(FutureWarning):
+                Index([], foo="bar")
```
Comment (jreback) [inline pandas/tests/indexes/base_class/test_constructors.py:34]: might be slightly tricky but can you check based on the index construction fixture so *all* klasses are convered for checking an in valid keyword shows the deprecation?

File: doc/source/whatsnew/v1.3.0.rst [hunk 4]
```diff
@@ -142,7 +142,7 @@ Other API changes
 
 Deprecations
 ~~~~~~~~~~~~
-
+- Deprecated allowing subclass-specific keyword arguments in the :class:`Index` constructor, use the specific subclass directly instead (:issue:`22315`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:145]: can you add all of the issues that you listed in the top of the PR here.

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. a testing request and some small comments. ping on green.

File: doc/source/whatsnew/v1.3.0.rst (hunk 4 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v1.3.0.rst:145]: its a little ambiguous, might want to not close until the deprecation is enforced?

File: pandas/tests/indexes/base_class/test_constructors.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/base_class/test_constructors.py:34]: None of the others accept kwargs, except for PeriodIndex which checks them (though #38615)

File: pandas/core/resample.py [hunk 5]
```diff
@@ -1946,6 +1946,10 @@ def _asfreq_compat(index, freq):
     new_index: Index
     if isinstance(index, PeriodIndex):
         new_index = index.asfreq(freq=freq)
-    else:
-        new_index = Index([], dtype=index.dtype, freq=freq, name=index.name)
+    if isinstance(index, DatetimeIndex):
+        new_index = DatetimeIndex([], dtype=index.dtype, freq=freq, name=index.name)
+    if isinstance(index, TimedeltaIndex):
```
Comment (topper-123) [inline pandas/core/resample.py:1951]: Shouldn´t this be an `elif`?

Comment (jbrockmendel) [reply pandas/core/resample.py:1951]: sure, either way works

Comment (jbrockmendel) [reply pandas/core/resample.py:1951]: yep, will update

File: doc/source/whatsnew/v1.3.0.rst (hunk 4 above)
Comment (jreback) [reply doc/source/whatsnew/v1.3.0.rst:145]: i would list them and we can close them with this PR. its not worth keeping them around now that they are deprecated. if you want to open a dedicated issues that lists the things that we want to check once this deprecation is enforced ok by me. (e.g. simply linking to those 4 issues is fine as well).

Comment (jbrockmendel) [reply doc/source/whatsnew/v1.3.0.rst]: updated + greenish

File: pandas/core/resample.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/resample.py:1952]: done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
can you merge master once more. i realized that the period keyword removal was merged *before* this

**Comment by jreback** [comment]
somethings still failing in precommit

**Comment by jreback** [comment]
nice!

**Pull request merged**